src/overlay/TCPPeer.h

```diff
@@ -26,15 +28,81 @@ class TCPPeer : public Peer
     static constexpr size_t BUFSZ = 0x40000; // 256KB
 
   private:
+    // Helper class which provides invariance for various data structures;
+    // Each method should be called from the same thread
+    class ThreadRestrictedVars
+    {
+        std::deque<TimestampedMessage> mWriteQueue;
+        std::vector<asio::const_buffer> mWriteBuffers;
+        bool const mUseBackgroundThread;
+        bool mWriting{false};
+        bool mSocketShutdownScheduled{false};
```

Comment:
It _looks_ from the way it's used like `mSocketShutdownScheduled` is now redundant with `mDropStarted`, but I might be misreading?